config/catalog.go

```diff
@@ -40,41 +58,102 @@ func (config *CatalogConfig) normalize(cofnigPath string) {
-func ReadCatalogConfig(opts *options.TerragruntOptions) (*TerragruntConfig, error) {
-	// We must first find the closest configuration from the current directory to ensure that it is not the root configuration,
-	// otherwise when we try to pull it via the include block it gets an error "Only one level of includes is allowed".
-	if !files.FileExists(opts.TerragruntConfigPath) {
-		fileToFind := filepath.Base(opts.TerragruntConfigPath)
-		if configPath, err := findInParentFolders([]string{fileToFind}, nil, opts); err != nil {
-			return nil, err
-		} else if configPath != "" {
-			opts.TerragruntConfigPath = configPath
-		}
+// ReadCatalogConfig looks for the "nearest" `terragrunt.hcl` in the parent directories if the given `opts.TerragruntConfigPath` does not exist. Since our normal parsing `ParseConfig` does not always work, as some `terragrunt.hcl` files are meant to be used from an `include` and/or they might use `find_in_parent_folders` and they only work from certain child folders, it parses this file to see if the config contains `include{...find_in_parent_folders()...}` block to determine if it is the root configuration. If it finds `terragrunt.hcl` that already has `include`, then read that configuration as is, oterwise generate a stub child `terragrunt.hcl` in memory with an `include` to pull in the one we found.
+func ReadCatalogConfig(parentCtx context.Context, opts *options.TerragruntOptions) (*CatalogConfig, error) {
+	configPath, configString, err := findCatalogConfig(parentCtx, opts)
+	if err != nil || configPath == "" {
+		return nil, err
 	}
+	opts.TerragruntConfigPath = configPath
 
-	configString, err := util.ReadFileAsString(opts.TerragruntConfigPath)
+	ctx := NewContext(parentCtx, opts)
+	ctx.ParserOptions = append(ctx.ParserOptions, hclparse.WithHaltOnErrorOnlyForBlocks([]string{MetadataCatalog}))
+	ctx.ConvertToTerragruntConfigFunc = convertToTerragruntCatalogConfig
+
+	config, err := ParseConfigString(ctx, configPath, configString, nil)
 	if err != nil {
 		return nil, err
 	}
 
-	// try to check if this is the root config, if yes then read the config as is.
-	if rootConfigReg.MatchString(configString) {
-		return ParseConfigString(configString, opts, nil, opts.TerragruntConfigPath, &EvalContextExtensions{})
+	return config.Catalog, nil
+}
+
+func findCatalogConfig(ctx context.Context, opts *options.TerragruntOptions) (string, string, error) {
+	var (
+		configPath        = opts.TerragruntConfigPath
+		configName        = filepath.Base(configPath)
+		catalogConfigPath string
+	)
+
+	for {
+		opts = &options.TerragruntOptions{
+			TerragruntConfigPath: filepath.Join(filepath.Dir(configPath), util.UniqueId(), configName),
+			MaxFoldersToCheck:    opts.MaxFoldersToCheck,
+		}
+
+		select {
+		case <-ctx.Done():
+			return "", "", nil
+		default:
+			// continue
+		}
```

Comment:
This allows you to stop the process by pressing Ctrl-C, in case the loop is endless, in case the functions of the `filepath` package do not work correctly under a certain operating system.